src/program/top/top.lua

```diff
@@ -4,152 +4,303 @@ module(..., package.seeall)
 
 local ffi = require("ffi")
 local C = ffi.C
+local S = require("syscall")
 local lib = require("core.lib")
 local shm = require("core.shm")
 local counter = require("core.counter")
-local S = require("syscall")
 local histogram = require("core.histogram")
 local usage = require("program.top.README_inc")
+local file = require("lib.stream.file")
+local fiber = require("lib.fibers.fiber")
+local sleep = require("lib.fibers.sleep")
+local inotify = require("lib.ptree.inotify")
+local op = require("lib.fibers.op")
+local cond = require("lib.fibers.cond")
+local channel = require("lib.fibers.channel")
 
-local long_opts = {
-   help = "h", list = "l"
-}
-
-function clearterm () io.write('\027[2J') end
+-- First, the global state for the app.
+--
+local snabb_state = { instances={}, counters={}, histograms={} }
+local ui = {
+   tree=nil, focus=nil, wake=cond.new(), rows=24, cols=80,
+   show_empty=false, show_rates=true,
+   show_links=true, show_apps=true, show_engine=true,
+   paused=false }
 
-function run (args)
-   local opt = {}
-   local object = nil
-   function opt.h (arg) print(usage) main.exit(1) end
-   function opt.l (arg) object = arg              end
-   args = lib.dogetopt(args, opt, "hl:", long_opts)
+local function needs_redisplay(keep_tree)
+   if not keep_tree then ui.tree = nil end
+   ui.wake:signal()
+end
 
-   if #args > 1 then print(usage) main.exit(1) end
-   local target_pid = select_snabb_instance(args[1])
+-- Now the core code that monitors /var/run/snabb and updates
+-- snabb_state.
+--
+local function is_dir(name)
+   local stat = S.lstat(name)
+   return stat and stat.isdir
```

Comment:
There's `lwutil.dir_exists`, which does the same thing (except it calls `stat` instead of `lstat`). So maybe is worth to move use that function or move it to `core/lib.lua`. Could be done in another PR.

Reply:
In this case we actually want to use lstat, so we can know what's a symbolic link and what's a directory.  It's a short function anyway...